### PR TITLE
lots more fixes and tweaks

### PR DESCRIPTION
**Pull request by adamdruppe**

### CODE REVIEW

File: app/controllers/enrollments_controller.rb [hunk 1]
```diff
@@ -177,9 +178,9 @@ def enrollment_submitted_crm_actions
       # actually pulled from the Contact and the API won't let us
       # access or update them through the CampaignMember.
       contact.Phone = @enrollment.phone
-      contact.City = @enrollment.city
-      contact.State = @enrollment.state
-      contact.save!
+      contact.OtherCity = @enrollment.city
```
Comment (sadleb) [inline app/controllers/enrollments_controller.rb]: So dumb that we have to do this.  Something seems wrong.

File: app/views/enrollments/_form.html.erb [hunk 2]
```diff
@@ -81,14 +81,16 @@
                   <%= f.text_field :state, 'class' => 'form-control short', 'placeholder' => 'State', 'style' => 'display: inline' %>
                 </div>
                 <div class="form-group student">
-                  <label>Your date of birth</label>
-                  <%= f.text_field :birthdate, 'class' => 'form-control short', 'placeholder' => '2/28/1994' %>
+                  <label>Your date of birth (MM/DD/YYYY)</label>
+                  <%= f.text_field :birthdate, 'class' => 'form-control short', 'placeholder' => '2/28/1994', 'pattern' => '[0-9]{1,2}/[0-9]{1,2}/[0-9]{4}' %>
```
Comment (sadleb) [inline app/views/enrollments/_form.html.erb]: that's an easy way to do it, nice!

File: app/views/enrollments/_form.html.erb [hunk 3]
```diff
@@ -81,14 +81,16 @@
...
                 </div>
 
+                <% if @student_id_required %>
                 <div class="form-group student">
                   <label>Your <%= @program_site %> Student ID</label>
-                  <%= f.text_field :student_id, 'class' => 'form-control short', 'title' => 'Student ID' %>
+                  <%= f.text_field :student_id, 'class' => 'form-control short', 'title' => 'Student ID', 'required' => 'required' %>
```
Comment (sadleb) [inline app/views/enrollments/_form.html.erb]: what happens if this is hidden, but still required?  it wouldn't be rendered so it wouldn't cause an error, right?

Comment (adamdruppe) [inline app/views/enrollments/_form.html.erb]: On Sat, Apr 04, 2015 at 08:03:56PM -0700, Brian Sadler wrote:
> what happens if this is hidden, but still required?  it wouldn't be rendered so it wouldn't cause an error, right?

correct, it doesn't get put out in the HTML at all so it
works. (You can confirm this by signing up as a coach, those
campaigns don't ask for student id yet you can still go through
the process. We'll want to actually do this on staging to confirm
but it works on my box now.)

File: app/views/enrollments/_form.html.erb [hunk 4]
```diff
@@ -140,7 +142,7 @@
                 <div class="form-group both">
                   <label class="student">Year of expected graduation</label>
                   <label class="coach">Graduation year (or expected graduation)</label>
-                  <%= f.text_field :anticipated_graduation, 'class' => 'form-control short', 'step' => '1', 'value' => '2017', 'type' => 'number', 'min' => '1900' %>
+                  <%= f.text_field :anticipated_graduation, 'class' => 'form-control short', 'step' => '1', 'type' => 'number', 'min' => '1900' %>
```
Comment (sadleb) [inline app/views/enrollments/_form.html.erb]: min 1900, ha!  geysers

File: lib/salesforce.rb [hunk 5]
```diff
@@ -51,6 +51,11 @@ def get_client
       client.sobject_module = SFDC_Models
       authenticate(client)
 
+      # We need to update this too so the SFDC_Models don't try to use
+      # an old client between server restarts - this will avoid the session
+      # expired problem as we reauthorize and fix the global too.
+      Databasedotcom::Sobject::Sobject.client = client
```
Comment (sadleb) [inline lib/salesforce.rb]: i think we talked about sometimes having to write code for x=x.  here is a prime example.

**Pull request merged**